library/std/src/sys/uefi/common.rs

```diff
@@ -0,0 +1,269 @@
+//! Contains most of the shared UEFI specific stuff. Some of this might be moved to `std::os::uefi`
+//! if needed but no point in adding extra public API when there is not Std support for UEFI in the
+//! first place
+
+use r_efi::efi::Guid;
+
+use crate::io::{self, const_io_error};
+use crate::mem::MaybeUninit;
+use crate::os::uefi;
+use crate::ptr::NonNull;
+
+// Locate handles with a particular protocol GUID
+/// Implemented using `EFI_BOOT_SERVICES.LocateHandles()`
+pub(crate) fn locate_handles(mut guid: Guid) -> io::Result<Vec<NonNull<crate::ffi::c_void>>> {
+    fn inner(
+        guid: &mut Guid,
+        boot_services: NonNull<r_efi::efi::BootServices>,
+        buf_size: &mut usize,
+        buf: *mut r_efi::efi::Handle,
+    ) -> io::Result<()> {
+        let r = unsafe {
+            ((*boot_services.as_ptr()).locate_handle)(
+                r_efi::efi::BY_PROTOCOL,
+                guid,
+                crate::ptr::null_mut(),
+                buf_size,
+                buf,
+            )
+        };
+
+        if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    let boot_services = boot_services();
+    let mut buf_len = 0usize;
+
+    match inner(&mut guid, boot_services, &mut buf_len, crate::ptr::null_mut()) {
+        Ok(()) => unreachable!(),
+        Err(e) => match e.kind() {
+            io::ErrorKind::FileTooLarge => {}
+            _ => return Err(e),
+        },
+    }
+
+    // The returned buf_len is in bytes
+    let mut buf: Vec<r_efi::efi::Handle> =
+        Vec::with_capacity(buf_len / crate::mem::size_of::<r_efi::efi::Handle>());
+    match inner(&mut guid, boot_services, &mut buf_len, buf.as_mut_ptr()) {
+        Ok(()) => {
+            // SAFETY: This is safe because the call will succeed only if buf_len >= required
+            // length. Also, on success, the `buf_len` is updated with the size of bufferv (in
+            // bytes) written
+            unsafe { buf.set_len(buf_len / crate::mem::size_of::<r_efi::efi::Handle>()) };
+            Ok(buf.into_iter().filter_map(|x| NonNull::new(x)).collect())
+        }
+        Err(e) => Err(e),
+    }
+}
+
+/// Open Protocol on a handle
+/// Implemented using `EFI_BOOT_SERVICES.OpenProtocol()`
+pub(crate) fn open_protocol<T>(
```

Comment:
After taking a closer look at open_protocol, none of the error cases can happen in this implementation. However, as far as I know, it is possible for UEFI implementations to have additional errors.

So what do you think @dvdhrm? Should I keep the returned Result in case an implementation wants to return some custom error in a weird edge case? Or just stick to the spec and remove the Result?